autogpt/llm/chat.py

```diff
@@ -28,233 +32,187 @@ def create_chat_message(role, content) -> Message:
-            logger.debug(f"Tokens remaining for response: {tokens_remaining}")
-            logger.debug("------------ CONTEXT SENT TO AI ---------------")
-            for message in current_context:
-                # Skip printing the prompt
-                if message["role"] == "system" and message["content"] == prompt:
-                    continue
-                logger.debug(f"{message['role'].capitalize()}: {message['content']}")
-                logger.debug("")
-            logger.debug("----------- END OF CONTEXT ----------------")
-            agent.log_cycle_handler.log_cycle(
-                agent.config.ai_name,
-                agent.created_at,
-                agent.cycle_count,
-                current_context,
-                CURRENT_CONTEXT_FILE_NAME,
-            )
+    # Add messages from the full message history until we reach the token limit
+    next_message_to_add_index = len(agent.history) - 1
+    insertion_index = len(message_chain)
+    # Count the currently used tokens
+    current_tokens_used = message_chain.token_length
+
+    # while current_tokens_used > 2500:
+    #     # remove memories until we are under 2500 tokens
+    #     relevant_memory = relevant_memory[:-1]
+    #     (
+    #         next_message_to_add_index,
+    #         current_tokens_used,
+    #         insertion_index,
+    #         current_context,
+    #     ) = generate_context(
+    #         prompt, relevant_memory, agent.history, model
+    #     )
+
+    # Account for user input (appended later)
+    user_input_msg = create_chat_message("user", user_input)
+    current_tokens_used += count_message_tokens([user_input_msg], model)
+
+    current_tokens_used += 500  # Reserve space for new_summary_message
+
+    # Add Messages until the token limit is reached or there are no more messages to add.
+    while next_message_to_add_index >= 0:
+        message_to_add = agent.history[next_message_to_add_index]
+
+        tokens_to_add = count_message_tokens([message_to_add], model)
+        if current_tokens_used + tokens_to_add > send_token_limit:
+            break
+
+        # Add the most recent message to the start of the chain,
+        #  after the system prompts.
+        message_chain.insert(insertion_index, agent.history[next_message_to_add_index])
+        current_tokens_used += tokens_to_add
+        next_message_to_add_index -= 1
+
+    # Update & add summary of trimmed messages
+    if len(agent.history) > 0:
+        new_summary_message, trimmed_messages = agent.history.trim_messages(
+            current_message_chain=list(message_chain),
+        )
+        tokens_to_add = count_message_tokens([new_summary_message], model)
+        message_chain.insert(insertion_index, new_summary_message)
+        current_tokens_used += tokens_to_add - 500
+
+        memory_store = get_memory(cfg)
+        for ai_msg, result_msg in agent.history.filter_message_pairs(trimmed_messages):
```

Comment:
That logic is wrong, you cannot rely on having only the (assistant, result) tuple in the summarized messages since we stop to shove in the frame context when we get over the token limit no matter if it is on an assistant message or a result message. That causes the out of bounds error. I would move the memory collection at the end of the agent loop without doing the tuple summarization and do that only when you need it. (Actually where do we need it?)

Reply:
Should be resolved in 20002e46dec82c4f5621fca61367f51654419baf